### PR TITLE
Handle encryption state in web interface

### PR DESCRIPTION
**Pull request by schiessle**
- [x] add a nice looking lock in front of a end-to-end encrypted folder
- [x] handle permissions correctly
- [x] hide '...' menu for encrypted folders

The easiest way to review is by setting "encrypted" in the filecache table to "1" for a folder and check if all the file actions are gone.

### CODE REVIEW

**Comment by schiessle** [comment]
@nextcloud/javascript could need some help from our Java Script experts to hide the complete "..." menu for end-to-end encrypted folders.

File: apps/files/js/fileactions.js [hunk 1]
```diff
@@ -370,6 +370,9 @@
 		_renderMenuTrigger: function($tr, context) {
 			// remove previous
 			$tr.find('.action-menu').remove();
+			var isE2EEncrypted = $tr.data('e2eencrypted');
+
+			if (isE2EEncrypted) return;
```
Comment (danxuliu) [inline apps/files/js/fileactions.js]: I think that the most common style in our JavaScript code is
```
if (isE2EEncrypted) {
    return;
}
```

File: core/js/files/client.js [hunk 2]
```diff
@@ -304,6 +309,13 @@
 				data.hasPreview = true;
 			}
 
+			var isEncryptedProp = props['{' + Client.NS_NEXTCLOUD + '}is-encrypted'];
+			if (!_.isUndefined(isEncryptedProp)) {
+				data.isEncrypted = isEncryptedProp === '1';
```
Comment (danxuliu) [inline core/js/files/client.js]: Just to be sure, the server is guaranteed to always return `'1'`, never `1`, `true` or `'true'`, right?

File: core/js/mimetypelist.js [hunk 3]
```diff
@@ -108,6 +108,7 @@ OC.MimeTypeList={
     "folder-public",
     "folder-shared",
     "folder-starred",
+    "folder-encrypted",
```
Comment (danxuliu) [inline core/js/mimetypelist.js]: Please amend to keep the alphabetical order :-)

**Review by danxuliu — COMMENTED** [review]
If you could add unit tests for [not rendering the actions menu trigger on encrypted folders](https://github.com/nextcloud/server/blob/handle-encryption-state-in-web-interface/apps/files/tests/js/fileactionsSpec.js#L112), [rendering encrypted folder icon for encrypted folders](https://github.com/nextcloud/server/blob/handle-encryption-state-in-web-interface/apps/files/tests/js/filelistSpec.js#L1441) and [ensuring that the _is-encrypted_ property is parsed as expected in a FileInfo](https://github.com/nextcloud/server/blob/handle-encryption-state-in-web-interface/core/js/tests/specs/files/clientSpec.js#L514) that would be great :-)

Besides my usual nitpicking the code looks good ;-)

File: core/js/files/client.js (hunk 2 above)
Comment (schiessle) [reply core/js/files/client.js]: yes, that's what's stored in the file cache table

**Comment by schiessle** [comment]
@danxuliu thanks for the review... I fixed the small stuff. May I ask you to have a look at the test? This would be awesome because I'm really bad with JS tests... Thanks a lot!

**Comment by icewind1991** [comment]
- [ ] I can still enter an encrypted folder
- [ ] Share icon should stay aligned when the dots aren't there

    image

Code looks good otherwise

**Comment by danxuliu** [comment]
@schiessle
>  May I ask you to have a look at the test?

No problem :-) I will need some clarification on the expected behaviour, though. What actions should be available on encrypted folders? And why are those actions available and not others?

I have also realized that just not rendering the three dots button is not a proper solution. Right now an encrypted folder can be favourited by clicking on the star icon on the left. However, [once the checkbox is placed where the favourite icon is now](https://github.com/nextcloud/server/pull/6709), the favourite action would be shown inside the three-dots menu, so the three-dots icon should be shown even for encrypted folders, although only with a single action.

So, in the same way that file actions are currently filtered based on permissions and mime types, an option could be to filter all actions on encrypted folders except those marked as _allowed-on-encrypted_ (or something like that). I could implement that if needed... although probably not in the next days, sorry :-S

@icewind1991
>  I can still enter an encrypted folder

If you should not enter an encrypted folder I would expect the server to prevent that. Otherwise, even if fixed in the web UI, you could simply replicate the current web UI behaviour "manually" and get access to the folder. Maybe the server does not prevent it because there is no proper encryption in place and it would not allow that on a real case (assuming that you just used the trick of setting `encrypted` to `1` in the `filecache` table, but I do not know if that is what you did :-) )?

In any case, what should be the expected behaviour? Not allowing to click on the folder? Showing a warning/error saying that it is encrypted and that it can not be accessed? Showing a dialog that ask for a decryption key to be able to enter?

>  Share icon should stay aligned when the dots aren't there

If the filtering I described above is implemented I think that the alignment should be fixed in a different pull request, as it could be seen as a general case of _"When there are no actions to be shown in the file actions menu the file actions trigger should not be shown, and the in-line actions should be kept aligned"_ (although that would not happen anyway once the pull request to change the placement of the checkbox is merged, so... I would say that this has low priority).

**Comment by schiessle** [comment]
@icewind1991 did you enabled the end2end encryption app on the server? If it is enabled access to the folder should be blocked

**Comment by schiessle** [comment]
@danxuliu thanks a lot for the tests! All operations on the encrypted folders should be removed. At the moment just the option to "star" the folder is still there which is fine.

**Comment by danxuliu** [comment]
I have realized that there is no need to add another field to file actions to specify whether they are allowed on encrypted folders or not to filter the file actions by that field as [I initially thought](https://github.com/nextcloud/server/pull/6670#issuecomment-336635055).

When installed in the server, the [E2E encryption app](https://github.com/nextcloud/end_to_end_encryption) blocks all the permissions in responses to PROPFIND requests, and file actions are currently filtered based on the permissions they need and the permissions that the user has on the file. For example, the download action needs read permissions, while the move action needs update permissions. Therefore, as encrypted files have no permissions (in the web UI) everything should simply work.

But it does not :-P The reason is that the JavaScript files client [sets the read permission on all files](https://github.com/nextcloud/server/blob/246a5a5750dd2b8f990797b2c6cc4270a9a5b59d/core/js/files/client.js#L322), no matter the permissions returned by the server. Thus all the actions that need read permissions are enabled for encrypted files even it they should not.

Besides that, currently every action require a permission, even if they should not. For example the favorite action requires the read permission, but there is no need to be able to read the file to mark or unmark it as a favourite. That is probably due to the read permission being always on and the fact that some code that works with file permissions currently [does not work as expected if the actions requires no permissions](https://github.com/nextcloud/server/blob/a8f1902b02d02cc2b7d624b3d7b288f8aea84fdc/apps/files/js/fileactions.js#L238) (as it would filter out the action even if it should not).

Finally, when selecting several files on the file list [the delete action is hidden if any of the files can not be deleted](https://github.com/nextcloud/server/blob/1f19a45a8815460b5c1e0c50c017e2671e87c58f/apps/files/js/filelist.js#L2825). However, the _Move or copy_ and the _Download_ actions are always shown even if some of the selected files do not support that action.

With all of the above my proposed solution would be to fix the handling of permissions in the file list and then handle encrypted files just like any other file, as the available actions will be automatically adjusted based on the permissions of encrypted files. Does this sound sane?

**Comment by schiessle** [comment]
@danxuliu yes, the webdav plugin sets already the permission to '0' if you can make the javascript to keep this, this would probably work.

**Comment by danxuliu** [comment]
I have rewritten the commit history to squash the unit tests to their related commits, extract the fixing of whitespace errors to its own commit, remove the commit that hid the three dots menu (as the _Favorite_ and _Details_ action could be performed on encrypted folders), and add some missing details (setting the folder icon too on `markFileAsShared`).

This pull request now requires #7046 and #7047 to be merged first; this pull request should be rebased on master once that happens, as there are some minor conflicts.

Also note that there is still a bug in this code: when the _Share_ tab is shown the file row loses part of its metadata, and if the _Share_ tab is loaded again (if it is simply opened there is no problem) then the encrypted folder will appear as a regular folder. For reference, the problem happens when [updating the file info model due to the `sharesChanged` event being triggered](https://github.com/nextcloud/server/blob/17a31a51c62e36500f59b3c2d385ab79258f6bb6/apps/files_sharing/js/share.js#L195), but I have not found the cause yet :-(

**Comment by schiessle** [comment]
@danxuliu all dependencies are now merged, can you rebase this PR and take care of the merge conflicts? Thanks!

**Comment by danxuliu** [comment]
>  @danxuliu all dependencies are now merged, can you rebase this PR and take care of the merge conflicts? Thanks!

Done ;-)

I have also fixed the bug [mentioned in my last comment](https://github.com/nextcloud/server/pull/6670#issuecomment-341510190); it was caused by [`elementToFile` not setting the `isEncrypted` attribute](https://github.com/nextcloud/server/pull/6670/commits/37d8d3d858d66ccfc9b9a32606a9448bc15f5960#diff-dafefd9519924bec4efa9f5af4af8e68R957).

When rebasing onto current master I dropped the commit that fixed the whitespace errors (they are now fixed in master) and split the commit to show the E2E folder icon in two, one to add the `e2eencrypted` attribute to the file list rows and another one just to show the E2E folder icon.

This pull request is now ready to be reviewed! :-)

**Comment by schiessle** [comment]
@danxuliu great. I will try it out! 😃

**Comment by schiessle** [comment]
tested it and it works really great! Thanks @danxuliu 

@nickvergessen @icewind1991 @rullzer can you please review it? Thanks!

**Review by rullzer — APPROVED** [review]

**Review by LukasReschke — APPROVED** [review]
🐘

**Pull request merged**